// z-utility/dist/index.js
const getInstructions = () => {
    return [
        { id: 1, name: '开锁功能指令配置', useType: '短按开锁键', useTypeId: '', },
        { id: 2, name: '关锁功能指令配置', useType: '短按关锁键', useTypeId: '', },
        { id: 3, name: '寻车功能指令配置', useType: '短按寻车键', useTypeId: '', },
        { id: 4, name: '尾箱功能指令配置', useType: '长按三秒尾箱键', useTypeId: '', },
        { id: 7, name: '升窗功能指令配置', useType: '长按7秒关锁键', useTypeId: '', },
        { id: 8, name: '降窗功能指令配置', useType: '长按7秒开锁键', useTypeId: '', },
    ];
}

const getOutputConfig = (customConfigs) => {

    const defaultConfigs = [
        {
            key: 'unlock', // 功能标识（便于维护）
            items: [
                { id: 1, name: '短按开锁键' },//输出次数1 输出时间500ms 输出间隔0
                { id: 2, name: '短按两次开锁键' },//输出次数2 输出时间500ms 输出间隔500ms
            ]
        },
        {
            key: 'lock',
            items: [
                { id: 1, name: '短按关锁键' },//输出次数1 输出时间500ms 输出间隔0
            ]
        },
        {
            key: 'findCar',
            items: [
                { id: 1, name: '短按寻车键' },//寻车键：输出次数1 输出时间500ms 输出间隔0; 关锁键:输出次数3 输出时间500 输出间隔1000ms
                { id: 2, name: '短按关锁键' },
            ]
        },
        {
            key: 'trunk',
            items: [
                { id: 1, name: '短按两次尾箱键' },//输出次数2 输出时间500ms 输出间隔1000ms
                { id: 2, name: '长按三秒尾箱键' },//输出次数1 输出时间3000ms 输出间隔0
            ]
        },
        {
            key: 'windowUp',
            items: [
                { id: 1, name: '长按7秒关锁键' },//输出次数为1 输出时间为7000ms 输出间隔0
            ]
        },
        {
            key: 'windowDown',
            items: [
                { id: 1, name: '长按7秒开锁键' },//输出次数为1 输出时间为7000ms 输出间隔0
            ]
        }
    ];


    const finalConfigs = customConfigs ? [...defaultConfigs, ...customConfigs] : defaultConfigs;

    return finalConfigs.map(config => config.items);
}


const getControlItems = () => {
    const baseItems = [
        {
            id: 1,
            name: '开锁',
            enabled: true,
            icon: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/unlock@2x.png',
            ative: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/colour-unlock.png',
            evt: 'handleUnlock'
        },
        {
            id: 2,
            name: '关锁',
            enabled: true,
            icon: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/lock@2x.png',
            ative: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/colour-lock.png',
            evt: 'handleLock'
        },
        {
            id: 3,
            name: '尾箱',
            enabled: true,
            icon: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/tail_box@2x.png',
            evt: 'handleOpenTrunk'
        },
        {
            id: 4,
            name: '寻车',
            enabled: true,
            icon: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/seek_car@2x.png',
            evt: 'handleFindCar'
        },
        {
            id: 5,
            name: '升窗',
            enabled: true,
            icon: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/rise.png',
            evt: 'handlRaiseTheWindow'
        },
        {
            id: 6,
            name: '降窗',
            enabled: true,
            icon: 'https://k1sw.wiselink.net.cn/img/app2.0/sjc/drop.png',
            evt: 'handleLowerTheWindow'
        },
    ];



    return baseItems;
}

const initVoltage = (dy) => {
    const thresholds = [4.0, 3.9, 3.8, 3.7, 3.6, 3.5, 3.4, 3.3, 3.2, 3.1];
    const scores = [100, 90, 80, 70, 60, 50, 40, 30, 20, 10];
    const index = thresholds.findIndex(threshold => dy >= threshold);
    return index !== -1 ? scores[index] : 0;
}


const getBatteryLevel = (voltage) => {
    const thresholds = [90, 80, 70, 60, 50, 40, 30, 20, 10, 5];
    const values = ['100', '90', '80', '70', '60', '50', '40', '30', '20', '10'];
    const index = thresholds.findIndex(threshold => voltage > threshold);
    return index !== -1 ? values[index] : '1';
}


const getParseHexDataObject = (hexString) => {
    if (!hexString || hexString.length !== 30) {
        // console.error('数据长度不正确，需为24（仅数据体）或32（完整帧）字符');
        return null;
    }

    let dataBodyHex = hexString;
    if (hexString.length === 32) {
        dataBodyHex = hexString.substr(4, 24);
    }

    const bytes = [];
    for (let i = 0; i < dataBodyHex.length; i += 2) {
        bytes.push(parseInt(dataBodyHex.substr(i, 2), 16));
    }

    const result = {};
    result.inductionEnable = bytes[0] === 1;
    result.inductionMode = bytes[0] === 1 ? true : false; // 兼容原有induction字段

    // D[1] ACC状态：0关，1开
    result.accStatus = bytes[1] === 1;

    // D[2] 锁状态：0关，1开（自动开关锁下发标记）
    result.lock = bytes[2] === 1; // 兼容原有lock字段

    // D[3] 3V断电剩余时间：0-255分钟，255=不断电/最大
    result.powerOffRemainTime = bytes[3];
    result.powerOffRemainTimeDesc = bytes[3] === 255 ? '不断电' : `${bytes[3]}分钟`;

    // D[4] 感应检测次数：0-255
    result.inductionCheckTimes = bytes[4];

    // D[5] 自动感应模式：0=操作后失效，1=一直有效，2=操作后失效
    result.autoInductionMode = bytes[5];
    result.autoInductionModeDesc = {
        0: '操作后失效',
        1: '一直有效',
        2: '操作后失效'
    }[bytes[5]] || '未知模式';

    // D[6] 蓝牙断开自动锁车配置：0关，1开
    result.bleDisconnectLock = bytes[6] === 1;

    // D[7] 位域解析（bit0~bit7）
    const d7 = bytes[7];
    result.carWashMode = (d7 & 0x01) === 1; // bit0：洗车模式 0关1开
    result.bleBroadcastMode = (d7 >> 1) & 0x01; // bit1：蓝牙广播模式（对应0x3D）
    result.modeType = (d7 >> 2) & 0x01; // bit2：0外置模式 1内置模式
    result.modeTypeDesc = (d7 >> 2) & 0x01 ? '内置模式' : '外置模式';
    result.oilCircuitStatus = (d7 >> 3) & 0x01; // bit3：油路状态 0开1关
    result.oilCircuitStatusDesc = (d7 >> 3) & 0x01 ? '油路关' : '油路开';
    result.lockWindowUp = (d7 >> 4) & 0x01; // bit4：锁车升窗 0不升1升
    result.serialBroadcast = (d7 >> 5) & 0x01 ? '关' : '开'; // bit5：串口输出广播 0开1关
    result.workMode = (d7 >> 6) & 0x01; // bit6：0正常模式 1网约车模式
    result.workModeDesc = (d7 >> 6) & 0x01 ? '网约车模式' : '正常模式';
    result.netCarControl = (d7 >> 7) & 0x01; // bit7：网约车模式是否可控制 0不可1可
    result.netCarControlDesc = (d7 >> 7) & 0x01 ? '可控制' : '不可控制';

    // D[8] 感应关锁信号值（原感应缓冲值，对应0x11）
    result.inductionLockSignal = bytes[8];

    // D[9] 感应门把手开关：0关1开
    result.inductionHandle = bytes[9] === 1;

    // D[10] 信号值：0-100（绝对值）
    result.signalValue = Math.abs(bytes[10]); // 确保绝对值

    // D[11] 感应开锁信号值（对应0x22）
    result.inductionUnlockSignal = bytes[11];

    // ===================== 扩展参数 D[12]~D[14]（兼容低版本） =====================
    // D[12] 电压值：0-255（120=12.0V），低版本无该位则为undefined
    if (bytes.length >= 13) {
        result.voltage = (bytes[12] / 10).toFixed(1) + 'V'; // 兼容原有voltage字段
        // 电池剩余电量计算（需补充getBatteryLevel和initVoltage方法）
        result.electric = getBatteryLevel(initVoltage((bytes[12] / 10).toFixed(1)));
    } else {
        result.voltage = '未知';
        result.electric = 0;
    }

    // D[13] 位域解析（bit0~bit5）
    if (bytes.length >= 14) {
        const d13 = bytes[13];
        result.alwaysPower = (d13 & 0x01) === 1; // bit0：常供电开关 0关1开
        result.startInductionEnable = (d13 >> 1) & 0x01 ? '失效' : '生效'; // bit1：启动状态感应 0生效1失效
        result.remoteInductionEnable = (d13 >> 2) & 0x01; // bit2：遥控感应开关是否有效
        result.keyWorkMode = (d13 >> 3) & 0x01; // bit3：钥匙工作模式
        result.keyAlwaysPower = (d13 >> 4) & 0x01; // bit4：钥匙一直供电 0关1开
        result.pairStatus = (d13 >> 5) & 0x01 ? '未配对' : '已配对'; // bit5：0已配对1未配对（兼容旧固件）
    } else {
        result.alwaysPower = false;
        result.startInductionEnable = '未知';
        result.remoteInductionEnable = false;
        result.keyWorkMode = 0;
        result.keyAlwaysPower = false;
        result.pairStatus = '未知';
    }

    // D[14] 位域解析（bit0~bit2：配对连接序号0-7；bit3：预留）
    if (bytes.length >= 15) {
        const d14 = bytes[14];
        result.pairConnectIndex = d14 & 0x07; // bit0~bit2（0-7）
        result.reservedBit3 = (d14 >> 3) & 0x01; // bit3：预留
    } else {
        result.pairConnectIndex = 0;
        result.reservedBit3 = 0;
    }

    // ===================== 原有兼容字段 =====================
    result.supply = bytes[3]; // 兼容原有supply字段（D[3]断电剩余时间）

    // console.log('解析结果:', result);
    return result;
}


const getInstructionMap = (sendCommand) => {
    return {
        1: { // 开锁键
            1: () => sendCommand(0x33, [0x33, 0x06, 0x01, 0x00, 0x00]), // 短按开锁键 
            2: () => sendCommand(0x33, [0x33, 0x06, 0x02, 0x06, 0x00])  // 短按两次开锁键 
        },
        2: { // 关锁键
            1: () => sendCommand(0x34, [0x34, 0x06, 0x01, 0x00, 0x00])  // 短按关锁键（原注释笔误修正）
        },
        3: { // 寻车键
            1: () => sendCommand(0x36, [0x36, 0x06, 0x01, 0x00, 0x00]), // 短按寻车键
            2: () => sendCommand(0x36, [0x34, 0x06, 0x03, 0x06, 0x00])  // 三按关锁键（原注释对应功能）
        },
        4: { // 尾箱键
            1: () => sendCommand(0x35, [0x35, 0x06, 0x02, 0x06, 0x00]), // 短按两次尾箱键
            2: () => sendCommand(0x35, [0x35, 0x1E, 0x01, 0x00, 0x00])  // 长按3秒尾箱键
        },
        5: { // 左中门
            1: () => sendCommand(0x50, [0x50, 0x06, 0x01, 0x00, 0x00]), // 短按左中门键
            2: () => sendCommand(0x50, [0x50, 0x1E, 0x01, 0x00, 0x00])  // 长按3秒左中门键
        },
        6: { // 右中门
            1: () => sendCommand(0x51, [0x51, 0x06, 0x01, 0x00, 0x00]), // 短按右中门键
            2: () => sendCommand(0x51, [0x51, 0x1E, 0x01, 0x00, 0x00])  // 长按3秒右中门键
        },
        7: { // 升窗
            1: () => sendCommand(0x52, [0x34, 0x46, 0x01, 0x00, 0x00])  // 长按7秒关锁键（升窗功能）
        },
        8: { // 降窗
            1: () => sendCommand(0x53, [0x33, 0x46, 0x01, 0x00, 0x00])  // 长按7秒开锁键（降窗功能）
        }
    };
}
module.exports = {
    getInstructions,//设置按键指令集合
    getOutputConfig,//获取输出配置列表
    getControlItems,//钥匙按钮常量
    getParseHexDataObject,//解析蓝牙广播值
    getInstructionMap,//获取指令映射表的函数
}
